Orange/widgets/visualize/owscatterplotgraph.py

```diff
@@ -598,480 +491,308 @@ def update_tooltip(self, modifiers):
-            color = QColor("#505050")
-            l_opts = dict(color=color, position=abs(int(rotate) - 0.85),
-                          rotateAxis=(1, 0), movable=True)
-            self.reg_line_item = InfiniteLine(
-                pos=QPointF(min_x, start_y), pen=pg.mkPen(color=color, width=1),
-                angle=angle, label="r = {:.2f}".format(rvalue), labelOpts=l_opts)
-            if rotate:
-                self.reg_line_item.label.angle = 180
-                self.reg_line_item.label.updateTransform()
-            self.plot_widget.addItem(self.reg_line_item)
-
-    def can_draw_density(self):
-        return self.domain is not None and \
-            self.attr_color is not None and \
-            self.attr_color.is_discrete and \
-            self.shown_x.is_continuous and \
-            self.shown_y.is_continuous
-
-    def should_draw_density(self):
-        return self.class_density and self.n_points > 1 and self.can_draw_density()
-
-    def can_draw_regresssion_line(self):
-        return self.domain is not None and \
-               self.shown_x.is_continuous and \
-               self.shown_y.is_continuous
-
-    def set_labels(self, axis, labels):
-        axis = self.plot_widget.getAxis(axis)
-        if labels:
-            ticks = [[(i, labels[i]) for i in range(len(labels))]]
-            axis.setTicks(ticks)
-        else:
-            axis.setTicks(None)
-
-    def set_axis_title(self, axis, title):
-        self.plot_widget.setLabel(axis=axis, text=title)
-
-    def compute_sizes(self):
-        self.master.Information.missing_size.clear()
-        if self.attr_size in [None, OWPlotGUI.SizeByOverlap]:
-            size_data = np.full((self.n_points,), self.point_width,
-                                dtype=float)
-        else:
-            size_data = \
-                self.MinShapeSize + \
-                self.scaled_data.get_column_view(self.attr_size)[0][self.valid_data] * \
-                self.point_width
-        nans = np.isnan(size_data)
-        if np.any(nans):
-            size_data[nans] = self.MinShapeSize - 2
-            self.master.Information.missing_size(self.attr_size)
-
-        # scale sizes because of overlaps
-        if self.attr_size == OWPlotGUI.SizeByOverlap:
-            size_data = np.multiply(size_data, self.overlap_factor)
-
-        return size_data
+        self.overlaps = [len(self.coord_to_id[xy]) for xy in zip(x, y)]
+        self.overlap_factor = [1 + log2(o) for o in self.overlaps]
+
+    def _jitter_data(self, x, seed=0):
+        random = np.random.RandomState(seed=seed)
+        off = self.jitter_size * (np.max(x) - np.min(x)) / 25
+        return x + random.uniform(-off, off, len(x))
+
+    # Sizes
+    def get_sizes(self):
+        size_column = self.master.get_size_data()
+        if size_column is None:
+            if self.master.attr_size == OWPlotGUI.SizeByOverlap:
+                self.update_colors()  # color points by most frequent color
```

Comment:
I don't like `update_coordinates`, `update_sizes`, `update_shapes`, `update_colors`, `update_labels` calling each other. I believe (hope?) that this is not violated anywhere, except here.

The logic of the separation between the widget and the plot is currently simple: if the widget calls `update_something`, the plot will respond by calling `get_something_data`. If it can also result in calling `get_somethingelse_data`, the whole logic of updating may again become difficult to follow.

If handling overlaps would be the responsibility of the plot - which I believe is not, it must find the way out of this without re-calling `update_colors`.

However, I believe that this is responsibility of the widget. What happens in this code here is that the sizes are provided by the plot instead of by the widget. If we want sizes by overlaps, it is the widget that needs to compute the overlaps and return the appropriate sizes.

Also, the widget, not the plot, must provide the colors.